src/diagnostics/MOM_diagnose_KdWork.F90

```diff
@@ -0,0 +1,68 @@
+!> Provides diagnostics of work due to a given diffusivity
+module MOM_diagnose_kdwork
+
+! This file is part of MOM6. See LICENSE.md for the license.
+
+use MOM_error_handler, only : MOM_error, FATAL, WARNING
+use MOM_grid,          only : ocean_grid_type
+use MOM_unit_scaling,  only : unit_scale_type
+use MOM_variables,     only : thermo_var_ptrs
+use MOM_verticalGrid,  only : verticalGrid_type
+
+implicit none ; private
+
+#include <MOM_memory.h>
+
+public diagnoseKdWork
+
+! A note on unit descriptions in comments: MOM6 uses units that can be rescaled for dimensional
+! consistency testing. These are noted in comments with units like Z, H, L, and T, along with
+! their mks counterparts with notation like "a velocity [Z T-1 ~> m s-1]".  If the units
+! vary with the Boussinesq approximation, the Boussinesq variant is given first.
+
+contains
+!> Diagnose the implied "work", or buoyancy forcing & its integral, due to a given diffusivity and column state.
+  subroutine diagnoseKdWork(G, GV, US, N2, Kd, Bdif_flx, dz, Bdif_flx_dz)
+  type(ocean_grid_type),   intent(in) :: G     !< Grid type
+  type(verticalGrid_type), intent(in) :: GV    !< ocean vertical grid structure
+  type(unit_scale_type),   intent(in) :: US    !< A dimensional unit scaling type
+  real, dimension(SZI_(G),SZJ_(G),SZK_(GV)+1), &
+       intent(in)                     :: N2    !< Buoyancy frequency [T-2 ~> s-2]
+  real, dimension(SZI_(G),SZJ_(G),SZK_(GV)+1), &
+       intent(in)                     :: Kd    !< Diffusivity [H2 T-1 ~> m2 s-1]
+  real, dimension(SZI_(G),SZJ_(G),SZK_(GV)+1), &
+       intent(out)                    :: Bdif_flx !< Buoyancy flux [H2 T-3 ~> m2 s-3]
+  real, dimension(SZI_(G),SZJ_(G),SZK_(GV)), &
+       intent(in), optional           :: dz    !< Grid spacing [H ~> m]
+  real, dimension(SZI_(G),SZJ_(G),SZK_(GV)), &
+       intent(out), optional          :: Bdif_flx_dz !< Buoyancy flux over layer [H3 T-3 ~> m3 s-3]
+
+  integer :: i, j, k
+
+  !$OMP parallel do default(shared)
+  do j=G%jsc,G%jec ; do i=G%isc,G%iec
+    Bdif_flx(i,j,1) = 0.0
+    Bdif_flx(i,j,GV%ke+1) = 0.0
+    do K=2,GV%ke
```

Comment:
This block of code would be more efficient if it were rearranged so that the K-loop is the outermost loop.  (In some cases, optimizing compilers would do this for us.)  Consider revising this code to swap the loop order.